### PR TITLE
Move external imports

### PR DESCRIPTION
**Pull request by interpotential**
This should solve #68 - it was written quite quickly, but for what I can see it does the job.

### CODE REVIEW

**Comment by holzhey** [comment]
Have same problem here... your commit solve it!

File: src/CSS.php [hunk 1]
```diff
@@ -66,6 +66,29 @@ public function setImportExtensions(array $extensions)
...
+     * Move any import statements to the top
+     *
+     * @param $content string Nearly finished CSS content
+     *
+     * @return string
+     */
+    protected function moveImportsToTop($content) {
+
+        if (preg_match_all('#@import[^;]+;#', $content, $matches)) {
+
```
Comment (matthiasmullie) [inline src/CSS.php]: Redundant newline

File: src/CSS.php [hunk 2]
```diff
@@ -66,6 +66,29 @@ public function setImportExtensions(array $extensions)
...
+            // remove from content
+            foreach($matches[0] as $import) {
```
Comment (matthiasmullie) [inline src/CSS.php]: According to PSR-2 coding standard, there should be a space between `foreach` and `(`

File: src/CSS.php [hunk 3]
```diff
@@ -66,6 +66,29 @@ public function setImportExtensions(array $extensions)
...
     /**
+     * Move any import statements to the top
+     *
+     * @param $content string Nearly finished CSS content
+     *
+     * @return string
+     */
+    protected function moveImportsToTop($content) {
+
+        if (preg_match_all('#@import[^;]+;#', $content, $matches)) {
```
Comment (matthiasmullie) [inline src/CSS.php]: Can you use `/` as a delimiter? Pet peave of mine - I wish everyone standardized on `/` (and that's what I already use elsewhere throughout the file) :)

File: src/CSS.php [hunk 4]
```diff
@@ -66,6 +66,29 @@ public function setImportExtensions(array $extensions)
     }
 
     /**
+     * Move any import statements to the top
+     *
+     * @param $content string Nearly finished CSS content
+     *
```
Comment (matthiasmullie) [inline src/CSS.php]: Redundant line between @param & @return

File: src/CSS.php [hunk 5]
```diff
@@ -66,6 +66,29 @@ public function setImportExtensions(array $extensions)
...
+     * @return string
+     */
+    protected function moveImportsToTop($content) {
+
+        if (preg_match_all('#@import[^;]+;#', $content, $matches)) {
+
+            // remove from content
+            foreach($matches[0] as $import) {
+                $content = str_replace($import, '', $content);
```
Comment (matthiasmullie) [inline src/CSS.php]: I don't think we should remove the @import occurences for specificity reasons.

Imagine this source code:

**a.css**
```css
body { color: red; }
```

**style.css**
```css
body { color: blue; }
@import url(a.css);
```

It would minify to:

**style.css**
```css
@import url(a.css);
body { color: blue; }
```

Because the import of a happens after style.css has already declared body color, the declaration on a.css would win (color would be red).
However, if we move the import to before the declaration, the last one defined would be blue (it would override red).

I think moving the assest imports up is a great idea. However, because of these rules with respect to specificity, I can't yet merge it as-is.
I think we can still move the imports to the beginning of the file, to make sure the browser starts loading them as soon as possible. However, we also need to keep them in place, to make sure the rules are not overwritten in another place. That would mean we have the imports in 2 places (at the top and in the original location), but the second one will probably just be read from cache anyway.

In short, my suggestion would be:

**style.css**
```css
@import url(a.css); /* This one to make sure browser starts loading quickly */
body { color: blue; }
@import url(a.css); /* This one to make sure it again overrides the above conflicting body color - it'll just be fetched from cache */
```


What do you think?

**Comment by matthiasmullie** [comment]
Hey @interpotential

Thanks for the pull request. It's a great idea!

I've added some feedback: most of it is coding standards nitpicking, but I do have a minor reservation about changing the location of the @import. I suggest to move it to the top of the file, but also keep it in place (see the other comment). What do you think?

File: src/CSS.php (hunk 5 above)
Comment (interpotential) [inline src/CSS.php]: You have a point, and moving around with the order of things in CSS is not my favourite kind of thing. However, browsers don't parse @import statements unless they precede all other statements. See: https://developer.mozilla.org/en-US/docs/Web/CSS/@import

In your example the text in the body would still be blue regardless of the second import statement, which simply gets ignored.

The only possible problem I'd see is, if you have a css file that looks like this:

**style.css**
```css
@import url(local.css);
@import url(//example.com/reset.css);
```
**local.css**
```css
body { color: green; }
```
**example.com/reset.css**
```css
body { color: blue; }
h1 { color: teal; }
```

In this case, body text would remain green with my pull request merged. But the alternative is worse; namely that text remains green, and the h1 color remains at the browser default as reset.css never gets read.

**Comment by interpotential** [comment]
Thanks for the coding style feedback - I've made the suggested adjustments. Looking forward to continuing the discussion on whether or not this is a good idea ;-).

**Comment by matthiasmullie** [comment]
I had no idea browsers ignore the `@import`s if they're not first.

It could still happen though - minifier also allows files to be combined & the 2nd file could have an import at the top, but combined it wouldn't be at the top anymore.
My suggestion, however, won't work because that `@import` would then also be ignored.
So while we could still run into some edge cases, this is definitely a big improvement already.

I'll merge it right away - thanks!
